### ISSUE TITLE
After upgrade from version 3.14.0 to 4.* running multiple test categories in parentheses separated with 'OR' stopped working

### ISSUE THREAD

**Issue by cjohnsson** [nunit/nunit#4651]
**Versions of lib**
.NET 6, NUnit 4.1.0 with NUnit3TestAdapter 4.5.0 and Microsoft.NET.Test.Sdk 17.9.0.

**Command to execute tests**
`dotnet test "testingnunit.dll" --verbosity d --filter "(TestCategory=TestsA&TestCategory!=TestsB)|(TestCategory=TestsC)"`

**How to reproduce**
1. Create a project called testingnunit with the provided example class in it. Build the project.
2. Run the test command, no tests will be found when running with NUnit version 4.0.0 or later.
Output is "**No test matches the given testcase filter**".
![image](https://github.com/nunit/nunit/assets/814161/ebf2a3c6-9436-490b-87fc-04e1d89cf946)
4. If you downgrade NUnit to version 3.14.0 and run the test command it finds the tests. TestInA and TestInC are exectued, TestInB is excluded as expected.
![image](https://github.com/nunit/nunit/assets/814161/0983212b-b40e-4c6c-8b12-8db873ee5fce)

**How to mitigate the problem when running with NUnit 4.1.0**
Run dotnet test for each filter within parentheses. 
First run `dotnet test "testingnunit.dll" --verbosity d --filter "(TestCategory=TestsA&TestCategory!=TestsB)"`, and then run `dotnet test "testingnunit.dll" --verbosity d --filter "(TestCategory=TestsC)"`

**Code example**
```csproj
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <TargetFramework>net6.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
    </PropertyGroup>

    <ItemGroup>
      <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
      <PackageReference Include="NUnit" Version="4.1.0" />
      <PackageReference Include="NUnit3TestAdapter" Version="4.5.0" />
    </ItemGroup>
</Project>
```

```csharp
using NUnit.Framework;

namespace TestingNunit;

[TestFixture, Category("TestsA")]
public class TestsA
{
    [Test]
    public void TestInA()
    {
        Assert.Pass();
    }
}

[TestFixture, Category("TestsB")]
public class TestsB
{
    [Test]
    public void TestInB()
    {
        Assert.Fail();
    }
}

[TestFixture, Category("TestsC")]
public class TestsC
{
    [Test]
    public void TestInC()
    {
        Assert.Pass();
    }
}
```

**Comment by cjohnsson** [comment]
The above is a problem if you have a project with many test categories and want to create complex filters.

In my case I want to execute all tests that are not ignored for a module, except the ones that are categorized as slow/obtrusive, and I also want to include not ignored cross-module "smoketests".
**My test filter:** 
(TestCategory=MyModule&TestCategory!=Obtrusive&TestCategory!=Slow&TestCategory!=Ignored)|(TestCategory=Smoketest&TestCategory!=Ignored)

**Comment by OsirisTerje** [comment]
Thanks!  So with 3.14.0 and adapter 4.5.0 it works, but with 4.1.0 and adapter 4.5.0 it doesn't work?

**Comment by cjohnsson** [comment]
> Thanks! So with 3.14.0 and adapter 4.5.0 it works, but with 4.1.0 and adapter 4.5.0 it doesn't work?

Exactly.

**Comment by varajan** [comment]
In my case even simple `--filter TestCategory=Smoke` returns `No test matches the given testcase filter`
- test agent: Linux-x64
- NUnit 4.1.0
- NUnit.Analyzers 4.2.0
- NUnit.Console 3.17.0
- NUnit3TestAdapter 4.5.0

And everything works fine with NUnit 3.13

**Comment by stevenaw** [comment]
I've pushed a repro here confirming the behaviour: https://github.com/nunit/nunit.issues/tree/main/Issue4651

A diff of the dump files for both framework `v3.14` and `v4.1.0` yield the same test filter:
```
<filter>
      <or>
         <and>
            <cat>TestsA</cat>
            <not>
               <cat>TestsB</cat>
            </not>
         </and>
         <cat>TestsC</cat>
      </or>
   </filter>
```

Yet no tests are found on 4.1.0

**Comment by stevenaw** [comment]
I've gotten closer on this one. I think I can commit to getting it done within 4.2

**Comment by stevenaw** [comment]
This has been fixed and should release as part of NUnit 4.2
